Improve .dockerignore compliance

mefyl opened the pull request:
Initially attempting to fix https://github.com/docker/compose/issues/3886, this PR revamps .dockerignore files to comply with the documentation on several aspects:

* AFAICT absolute paths where not handled correctly, although there was an attempt - confirmed manually. The issue is that leading slashes where handled before exceptions (patterns starting with "!").
* Wildcards / globbing where not handled in exceptions.
* The last matching line now takes precedence, instead of the previous "inclusion over exclusion" rule.

The codebase was also made greatly smaller / simpler.

Unit test for "should_check_directory", an internal helper, were dropped since the implementation changed. All other unit test are honored as-is, except one about absolute path whose expectation where, I think, wrong - path are never rooted in the host filesystem at this point.

GordonTheTurtle commented:
<!-- AUTOMATED:POULE:DCO-EXPLANATION -->
Please sign your commits following these rules:
https://github.com/moby/moby/blob/master/CONTRIBUTING.md#sign-your-work
The easiest way to do this is to amend the last commit:
~~~console
$ git clone -b "master" git@github.com:mefyl/docker-py.git somewhere
$ cd somewhere
$ git rebase -i HEAD~842354662816
editor opens
change each 'pick' to 'edit'
save the file and quit
$ git commit --amend -s --no-edit
$ git rebase --continue # and repeat the amend for each commit
$ git push -f
~~~

Amending updates the existing PR. You **DO NOT** need to open a new one.

shin- commented on docker/utils/build.py:
```diff
@@ -23,127 +27,90 @@ def exclude_paths(root, patterns, dockerfile=None):
+        while i < len(splitted):
+            if splitted[i] == '..':
+                del splitted[i]
+                if i > 0:
```
Does this mean that `../xxx` becomes `xxx`? If so, does it match the behavior of the engine?

shin- replied:
More generally, having some tests for `..` components would be nice

mefyl replied:
The official documentation about this is rather elusive. It does however stipulate that absolute paths are equivalent to relative paths, hence `/../foo` should be equivalent to `../foo`. It also stipulates that paths are run through Go's `filepath.Clean`, which explicitely makes this equivalence: ```Eliminate .. elements that begin a rooted path:
   that is, replace "/.." by "/" at the beginning of a path
```

From this, I would argue that yes, the specification prompts for ../foo to be equivalent to foo.

But as you smartly suggested, I tested with Docker `17.09.1-ce`, and `../foo` does not ignore a `foo` file at the root - I suppose it won't ignore anything ever.

However, brace yourself, if you use `/../foo`, it *does* ignore a `foo` file at the root, which violates the statements that absolute and relative paths are equivalent. Long story short: it's a mess.

It might be more reasonable to mimic the current engine behavior, although not following the spec could seem worse. The spec would probably need some updating if we choose to mimic the engine.

shin- replied:
Ok - it probably doesn't see a lot of use anyway. I say we keep it as is and we can always revisit it later if it's an issue.

mefyl replied:
I agree that even any `..` in a `.dockerignore` file doesn't make much sense to me. Especially since `anything/..` will be dropped, even if `anything` does not exist as a directory, so I really don't see the point. IMO we should even deprecate and then reject them.

shin- commented on docker/utils/build.py:
```diff
@@ -23,127 +27,90 @@ def exclude_paths(root, patterns, dockerfile=None):
+        # components are not relevant either, even if the whole
+        # pattern is only ".", as the Docker reference states: "For
+        # historical reasons, the pattern . is ignored."
+        splitted = [pt for pt in re.split(_SEP, p) if pt and pt != '.']
```
nit: `split` (or better yet, `split_path`)

shin- commented on docker/utils/build.py:
```diff
@@ -23,127 +27,90 @@ def exclude_paths(root, patterns, dockerfile=None):
+        if p.startswith('!') else
+        (False, normalize(p))
+        for p in patterns)
+    patterns = list(reversed(list(chain(
```
does `patterns` need to be a list? It looks like `walk` only ever uses it as an iterator.

mefyl replied:
Nah, there's a `for` loop in `walk` reusing the patterns. As for the intermediate `list` inside `reverse`, sadly python2's `reverse` expects a plain collection -_-

shin- commented on docker/utils/build.py:
```diff
@@ -23,127 +27,90 @@ def exclude_paths(root, patterns, dockerfile=None):
-        path_components = path.split(os.path.sep)
-    return fnmatch('/'.join(path_components), '/'.join(pattern_components))
+    def match(p):
+        if p[1][0] == '**':
```
We're using a modified `fnmatch` implementation that _should_ handle `**` patterns - did you find that not to be the case? If so, maybe we can get rid of the `utils/fnmatch.py` file as well and use Python's `fnmatch` module instead.

mefyl replied:
The previous implementation did not handle `**` in exceptions, and it didn't look easy with this approach while preserving the optimization to not explore everything recursively when we're sure a directory is entirely excluded. Also, the current implementation looks much shorter and simpler to me, leveraging the stdandard fnmatch - provided I actually use the standard fnmatch, granted ;)

shin- replied:
Cool, I'm fine with the new implementation if it works better than the old one :)
Can we remove utils/fnmatch.py from the tree in this case?

mefyl commented:
I don't really like the idea of squashing: I don't see what it brings to the table as you can always review the diffs for the whole merge, but I think it removes a lot. As in, readibility, and for instance I added two optional "optimization" commits that can easily be blamed back, and reverted individually while conserving the semantics if those optimisations cause problem.

Now, I don't mean to come in the project on my high horses and impose my ways, if you prefer I'll squash for now, but that's a discussion I'd like to have at some point.

shin- commented:
Yeah, I think it's fine, these commits are all individually clean and self-contained. 👍

DxCx commented:
Not intentionally :)
Ill try importing both packages and see if the functions around here returns different values.

Thanks that was very helpful, hope to provide with more info soon :)
